### PR TITLE
Feature/accounts

### PR DESCRIPTION
**Pull request by rishsingh07**
Payment History  added with filters.

- [x] Settings button
- [x] Responsive design
- [x] Show illustration and message when there are no donations 
- [x] Settings button

### CODE REVIEW

File: src/features/user/Account/components/PaymentRecord.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import { Modal } from '@material-ui/core';
+import React, { ReactElement } from 'react';
+import BackButton from '../../../../../out/assets/images/icons/BackButton';
```
Comment (norbertschuler) [inline src/features/user/Account/components/PaymentRecord.tsx:3]: Do no import stuff from the `out` folder which is generated while deploying. Using `public` instead would be right. Using `out` shows an error if running this locally with `npm run dev`.

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: src/features/user/Account/screens/index.tsx [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+                </div>
+              ) : (
+                paymentHistory?.items?.map((item, index) => {
+                  return <PaymentRecord record={item} index={index} />;
+                })
+              )}
+            </div>
+            <div className={styles.filterContainerDesktop}>
+              <div className={styles.filterHead}>
+                <p className={styles.filterTitle}>{t('filters')}</p>
```
Comment (norbertschuler) [inline src/features/user/Account/screens/index.tsx:205]: This resource is missing.

File: src/features/user/Account/screens/index.tsx [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  return Object.entries(props.accountingFilters).map((item) => {
+    return (
+      <button
+        className={`${styles.multiSelectInput} ${
+          props.filter === item[0] ? styles.multiSelectInputCheckTrue : ''
+        }`}
+        key={item[0]}
+        onClick={() => props.handleSetFilter(item[0])}
+      >
+        {item[0]}
```
Comment (norbertschuler) [inline src/features/user/Account/screens/index.tsx:256]: These items have to translations:

Bildschirmfoto 2021-04-05 um 12 26 46

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: src/features/user/Account/screens/index.tsx (hunk 3 above)
Comment (sarveshpro) [reply src/features/user/Account/screens/index.tsx:278]: Those values are coming from backend

File: src/features/user/Account/screens/index.tsx (hunk 2 above)
Comment (sarveshpro) [reply src/features/user/Account/screens/index.tsx]: added

**Review by norbertschuler — COMMENTED** [review]
- [x] `public/assets/images/icons/TransactionsNotFound.tsx` is 941kb and raising this message running locally with `npm run dev`:

```
[BABEL] Note: The code generator has deoptimised the styling of /Users/norbertschuler/Development/PFTP/planet-webapp/public/assets/images/icons/TransactionsNotFound.tsx as it exceeds the max of 500KB.
```

Should we perhaps reduce the complexity of this image avoiding adding 1MB more to download for the users?

**Comment by norbertschuler** [comment]
I viewed the result of the API call  GET https://app-development.plant-for-the-planet.org/app/paymentHistory with `x-locale: de` set, but still returning much data not formatted for German language but in English:
- I don't think the backend should do any localization work, but rather return language independent data to be formatted by the frontend 
- the API call is returning multiple localized resources e.g. strings, dates or numbers combined with currencies with cannot easily be reformatted in the frontend:

```json
{
	"items": [
		{
			"caption": "50 Tree Donation",   <-- not localized number combined with none translated text
			"reference": "000559221",
			"amount": 100,
			"currency": "EUR",
			"status": "success",
			"created": "2021-04-07 08:27:43",
			"details": [
				{
					"key": "amount",
					"caption": "Amount",  <-- none translated text
					"text": "EUR 100.00"  <-- not localized number combined with text
				},
				{
					"key": "created",
					"caption": "Paid Date", <-- none translated text
					"text": "04/07/21"      <-- not localized date
				},
				{
					"key": "project",
					"caption": "Project",   <-- none translated text
					"text": "Plant-for-Ethiopia"
				},
				{
					"key": "provider",
					"caption": "Provider",  <-- none translated text
					"text": "card"
				},
				{
					"key": "treeCost",
					"caption": "Tree Cost", <-- none translated text
					"text": "2.00"          <-- not localized number
				},
				{
					"key": "treeCount",
					"caption": "Tree Count", <-- none translated text
					"text": "50"             <-- not localized number
				},
				{
					"key": "lastUpdate",
					"caption": "Last Updated", <-- none translated text
					"text": "04/07/21"         <-- not localized date
				},
				{
					"key": "donorCertificate",
					"caption": "Tree certificate", <-- none translated text
					"text": "<a href=\"http://app-development.plant-for-the-planet.org/certificate/000559221\">detail.donor_certificate</a>"  
					^-- HTML from backend with missing translation
				}
			]
		}
	],
	"total": 1,
	"count": 1,
	"_links": {
		"self": "/app/paymentHistory?page=1",  <-- none translated text
		"first": "/app/paymentHistory?page=1", <-- none translated text
		"last": "/app/paymentHistory?page=1"   <-- none translated text
	},
	"_filters": {
		"Donations": "/app/paymentHistory?filter=donations",     <-- none translated text
		"In Progress": "/app/paymentHistory?filter=in-progress", <-- none translated text
		"Cancelled": "/app/paymentHistory?filter=cancelled",     <-- none translated text
		"Tree Cash": "/app/paymentHistory?filter=tree-cash"      <-- none translated text
	}
}
```

**Comment by sarveshpro** [comment]
> I viewed the result of the API call GET https://app-development.plant-for-the-planet.org/app/paymentHistory with `x-locale: de` set, but still returning much data not formatted for German language but in English:
> 
> * I don't think the backend should do any localization work, but rather return language independent data to be formatted by the frontend
> * the API call is returning multiple localized resources e.g. strings, dates or numbers combined with currencies with cannot easily be reformatted in the frontend:
> 
> ```json
> {
> 	"items": [
> 		{
> 			"caption": "50 Tree Donation",   <-- not localized number combined with none translated text
> 			"reference": "000559221",
> 			"amount": 100,
> 			"currency": "EUR",
> 			"status": "success",
> 			"created": "2021-04-07 08:27:43",
> 			"details": [
> 				{
> 					"key": "amount",
> 					"caption": "Amount",  <-- none translated text
> 					"text": "EUR 100.00"  <-- not localized number combined with text
> 				},
> 				{
> 					"key": "created",
> 					"caption": "Paid Date", <-- none translated text
> 					"text": "04/07/21"      <-- not localized date
> 				},
> 				{
> 					"key": "project",
> 					"caption": "Project",   <-- none translated text
> 					"text": "Plant-for-Ethiopia"
> 				},
> 				{
> 					"key": "provider",
> 					"caption": "Provider",  <-- none translated text
> 					"text": "card"
> 				},
> 				{
> 					"key": "treeCost",
> 					"caption": "Tree Cost", <-- none translated text
> 					"text": "2.00"          <-- not localized number
> 				},
> 				{
> 					"key": "treeCount",
> 					"caption": "Tree Count", <-- none translated text
> 					"text": "50"             <-- not localized number
> 				},
> 				{
> 					"key": "lastUpdate",
> 					"caption": "Last Updated", <-- none translated text
> 					"text": "04/07/21"         <-- not localized date
> 				},
> 				{
> 					"key": "donorCertificate",
> 					"caption": "Tree certificate", <-- none translated text
> 					"text": "<a href=\"http://app-development.plant-for-the-planet.org/certificate/000559221\">detail.donor_certificate</a>"  
> 					^-- HTML from backend with missing translation
> 				}
> 			]
> 		}
> 	],
> 	"total": 1,
> 	"count": 1,
> 	"_links": {
> 		"self": "/app/paymentHistory?page=1",  <-- none translated text
> 		"first": "/app/paymentHistory?page=1", <-- none translated text
> 		"last": "/app/paymentHistory?page=1"   <-- none translated text
> 	},
> 	"_filters": {
> 		"Donations": "/app/paymentHistory?filter=donations",     <-- none translated text
> 		"In Progress": "/app/paymentHistory?filter=in-progress", <-- none translated text
> 		"Cancelled": "/app/paymentHistory?filter=cancelled",     <-- none translated text
> 		"Tree Cash": "/app/paymentHistory?filter=tree-cash"      <-- none translated text
> 	}
> }
> ```



> I viewed the result of the API call GET https://app-development.plant-for-the-planet.org/app/paymentHistory with `x-locale: de` set, but still returning much data not formatted for German language but in English:
> 
> * I don't think the backend should do any localization work, but rather return language independent data to be formatted by the frontend
> * the API call is returning multiple localized resources e.g. strings, dates or numbers combined with currencies with cannot easily be reformatted in the frontend:
> 
> ```json
> {
> 	"items": [
> 		{
> 			"caption": "50 Tree Donation",   <-- not localized number combined with none translated text
> 			"reference": "000559221",
> 			"amount": 100,
> 			"currency": "EUR",
> 			"status": "success",
> 			"created": "2021-04-07 08:27:43",
> 			"details": [
> 				{
> 					"key": "amount",
> 					"caption": "Amount",  <-- none translated text
> 					"text": "EUR 100.00"  <-- not localized number combined with text
> 				},
> 				{
> 					"key": "created",
> 					"caption": "Paid Date", <-- none translated text
> 					"text": "04/07/21"      <-- not localized date
> 				},
> 				{
> 					"key": "project",
> 					"caption": "Project",   <-- none translated text
> 					"text": "Plant-for-Ethiopia"
> 				},
> 				{
> 					"key": "provider",
> 					"caption": "Provider",  <-- none translated text
> 					"text": "card"
> 				},
> 				{
> 					"key": "treeCost",
> 					"caption": "Tree Cost", <-- none translated text
> 					"text": "2.00"          <-- not localized number
> 				},
> 				{
> 					"key": "treeCount",
> 					"caption": "Tree Count", <-- none translated text
> 					"text": "50"             <-- not localized number
> 				},
> 				{
> 					"key": "lastUpdate",
> 					"caption": "Last Updated", <-- none translated text
> 					"text": "04/07/21"         <-- not localized date
> 				},
> 				{
> 					"key": "donorCertificate",
> 					"caption": "Tree certificate", <-- none translated text
> 					"text": "<a href=\"http://app-development.plant-for-the-planet.org/certificate/000559221\">detail.donor_certificate</a>"  
> 					^-- HTML from backend with missing translation
> 				}
> 			]
> 		}
> 	],
> 	"total": 1,
> 	"count": 1,
> 	"_links": {
> 		"self": "/app/paymentHistory?page=1",  <-- none translated text
> 		"first": "/app/paymentHistory?page=1", <-- none translated text
> 		"last": "/app/paymentHistory?page=1"   <-- none translated text
> 	},
> 	"_filters": {
> 		"Donations": "/app/paymentHistory?filter=donations",     <-- none translated text
> 		"In Progress": "/app/paymentHistory?filter=in-progress", <-- none translated text
> 		"Cancelled": "/app/paymentHistory?filter=cancelled",     <-- none translated text
> 		"Tree Cash": "/app/paymentHistory?filter=tree-cash"      <-- none translated text
> 	}
> }
> ```

Perhaps because translations are not yet added in the backend. we did this maybe because we were not sure of what transaction details we will show here. and adding new fields won't require an update on the frontend.

**Comment by sarveshpro** [comment]
> `public/assets/images/icons/TransactionsNotFound.tsx` is 941kb and raising this message running locally with `npm run dev`:
> 
> ```
> [BABEL] Note: The code generator has deoptimised the styling of /Users/norbertschuler/Development/PFTP/planet-webapp/public/assets/images/icons/TransactionsNotFound.tsx as it exceeds the max of 500KB.
> ```
> 
> Should we perhaps reduce the complexity of this image avoiding adding 1MB more to download for the users?

I agree, have asked Harsh to share a smaller version

**Comment by norbertschuler** [comment]
> Perhaps because translations are not yet added in the backend. we did this maybe because we were not sure of what transaction details we will show here. and adding new fields won't require an update on the frontend.

The backend project has not updated its translation resources from LingoHub for quite some time now. I created a [pull request](https://translate.lingohub.com/plant-for-the-planet/dashboard/planet-app-(backend)/
) to test if an update is still possible, but some languages have a lot of resources missing or not approved.

**Comment by sarveshpro** [comment]
wait for backend change to test

**Comment by norbertschuler** [comment]
- This does not yet work on staging as probably the backend API is not yet updated and thus not compatible with this pull requests yet.

**Comment by norbertschuler** [comment]
> wait for backend change to test

@sarveshpro Backend changes are released to staging, but not yet compatible with your code. I think the int values for the dates given are not what you expected: 
```
RangeError: Invalid time value
    at t.default (index.js:392)
    at l (getFormattedDate.ts:8)
    at w (PaymentRecord.tsx:166)
```

Example response now:
```json
{
	"items": [
		{
			"type": "tree-donation",
			"reference": "000679316",
			"amount": 20,
			"treeCount": 10.0,
			"currency": "EUR",
			"status": "success",
			"created": "2021-04-09T10:40:34+0000",
			"lastUpdate": "2021-04-09T10:40:58+0000",
			"details": {
				"amount": {
					"type": "amount",
					"amount": 2000,
					"currency": "EUR"
				},
				"created": {
					"date": "1617964834",
					"type": "date"
				},
				"project": {
					"text": "Plant-for-Ethiopia",
					"type": "text",
					"translate": false
				},
				"provider": {
					"text": "stripe_pi",
					"type": "text",
					"translate": false
				},
				"treeCost": {
					"type": "number",
					"format": "%.2f",
					"number": 2
				},
				"treeCount": {
					"type": "number",
					"format": "%d",
					"number": 10
				},
				"lastUpdate": {
					"date": "1617964857",
					"type": "date"
				},
				"donorCertificate": {
					"url": "http://app-staging.plant-for-the-planet.org/certificate/000679316",
					"text": "detail.donor_certificate",
					"type": "url",
					"translate": true
				}
			}
		}
	],
	"total": 1,
	"count": 1,
	"_links": {
		"self": "/app/paymentHistory?page=1",
		"first": "/app/paymentHistory?page=1",
		"last": "/app/paymentHistory?page=1"
	},
	"_filters": {
		"donations": "/app/paymentHistory?filter=donations",
		"in-progress": "/app/paymentHistory?filter=in-progress",
		"cancelled": "/app/paymentHistory?filter=cancelled",
		"tree-cash": "/app/paymentHistory?filter=tree-cash"
	}
}
```
Not sure why @jmiridis did not return dates like for `created` and `lastUpdate`?

**Comment by sagararyal** [comment]
PR doesn't handle 401s
Screen Shot 2021-04-19 at 3 20 51 PM

**Comment by norbertschuler** [comment]
Looking quite good, but I found some minor issues:

- [ ] Some translations for filters (e.g. in language as french) might be longer and do not have any left and right margin if using up all the space available?
- [ ] The payment method has not translation and also only uses lowercase letters. That does not look so good in other languages as English.
- [ ] The download link is always given for production and not using the same domain as the API call is using.

Bildschirmfoto 2021-04-22 um 15 11 35

**Comment by sarveshpro** [comment]
> Looking quite good, but I found some minor issues:
> 
> * [ ]  Some translations for filters (e.g. in languages as french) might be longer and do not have any left and right margin if using up all the space available?
> * [ ]  The payment method has no translation and also only uses lowercase letters. That does not look so good in other languages as English. Examples: `stripe - card  ` or `tree-cash - tree-cash` or `paypal - paypal `
> * [ ]  The download link is always given for production and not using the same domain as the API call is using.
> * [ ]  The returned value for `taxReceipt` (only for some items) is not visualized as any link, is that not yet wanted? (All the example links I got for my test accounts returned 404 errors.)
> 
> Bildschirmfoto 2021-04-22 um 15 11 35

- the buttons won't look nice in grid layout on desktop if they have variable sizes, should I change it to flex then? is any text going out of the buttons?
- URLs are returned by the backend, will add this to the backend channel
- should payment methods be translated? I can create resources for individual words like - stripe, card and tree-cash

**Comment by norbertschuler** [comment]
> * the buttons won't look nice in grid layout on desktop if they have variable sizes, should I change it to flex then? is any text going out of the buttons?

On a smaller screen it breaks the layout, otherwise only some padding is missing:
Bildschirmfoto 2021-04-23 um 08 04 00
Bildschirmfoto 2021-04-23 um 08 03 35

**Comment by norbertschuler** [comment]
> * should payment methods be translated?

As long as we do not know a complete list of all possible response value for payment methods, the client can not do the translation. Also the current responses e.g. `stripe - card` or `tree-cash - tree-cash or `paypal - paypal` are not really suitable as translation resource keys. So either the the backend returns the translated payment method string or a translation key and a complete list of possible response values to enable translations on client site.

**Comment by sarveshpro** [comment]
Please get this merged, account history will be hidden from users for now, but is accessible if viewed `/account/history` while being logged in. the components in this feature are required to build another feature(recurring donations, treemapper, common account header etc). so this code should go in develop.

**Comment by norbertschuler** [comment]
> Please get this merged, account history will be hidden from users for now, but is accessible if viewed `/account/history` while being logged in. the components in this feature are required to build another feature(recurring donations, treemapper, common account header etc). so this code should go in develop.

Ok, I put the remaining issues into https://github.com/Plant-for-the-Planet-org/planet-webapp/issues/1026 so we do not forget about them if we merge this pull request.

**Review by norbertschuler — APPROVED** [review]
Approving as requested. Remaining issue can be done in another pull request.

**Pull request merged**